.circleci/Dockerfile

```diff
@@ -0,0 +1,77 @@
+FROM debian:jessie as itms_installer
+
+ENV ITMS_VERSION "1.9.7"
+USER root
+
+WORKDIR /tmp
+
+# Install iTMSTransporter
+ADD https://github.com/janpio/fastlane-ubuntu-foo/raw/master/iTMSTransporter_installer_linux_$ITMS_VERSION.sh .
+RUN chmod +rx ./iTMSTransporter_installer_linux_$ITMS_VERSION.sh \
+	&& ./iTMSTransporter_installer_linux_$ITMS_VERSION.sh --target itms --noexec \
+	&& cd itms \
+	&& sed -e '15,43d' -i install_script.sh \
+	&& apt-get update \
+	&& apt-get install sudo \
+	&& ./install_script.sh
+
+FROM buildpack-deps:jessie as xar_builder
+
+ENV XAR_VERSION "1.6.1"
+USER root
+
+WORKDIR /tmp
+
+# Build xar
+ADD https://github.com/downloads/mackyle/xar/xar-$XAR_VERSION.tar.gz .
+RUN tar -xzf xar-$XAR_VERSION.tar.gz \
+	&& mv xar-$XAR_VERSION xar \
+	&& cd xar \
+	&& ./autogen.sh --noconfigure \
+	&& ./configure \
+	&& make 
+
+
+###############
+# Final image #
+###############
+FROM circleci/ruby:2.3
+MAINTAINER milch
+
+ENV PATH $PATH:/usr/local/itms/bin
+
+# Java versions to be installed
+ENV JAVA_VERSION 8u131
+ENV JAVA_DEBIAN_VERSION 8u131-b11-1~bpo8+1
+ENV CA_CERTIFICATES_JAVA_VERSION 20161107~bpo8+1
+
+# Needed for fastlane to work
+ENV LANG C.UTF-8
+ENV LC_ALL C.UTF-8
+
+# Required for iTMSTransporter to find Java
+ENV JAVA_HOME /usr/lib/jvm/java-8-openjdk-amd64/jre
+
+COPY --from=itms_installer /usr/local/itms /usr/local/itms
```

Comment:
Is this necessary? Really running the iTMS installer should automatically move the files there.

Reply:
Each `FROM` line in the Dockerfile essentially creates separate VMs (called stages), so this line is necessary to copy over the folder from the other stage.

Having separate "stages" is generally a good idea because it has implications on final image size and the layer cache. Each line in the Dockerfile for each of the stages is essentially a separate layer in the virtual file system. If this was all a single stage, updating iTMS would mean that essentially: 

1. The whole cache would need to be discarded
    - XAR would also need to be rebuilt 
2. Most of the final image would change completely, which would mean uploading hundreds of MB on a rebuild

It is generally a good idea to keep the amount of layers to a minimum - which is also why you see the pattern of having one `RUN` statement with every command joined by `&&` as opposed to having them in separate `RUN` statements. Anything that you clean up within one such statement won't result in increased image size. 

It also saves some space to discard any temporary directories or dependencies that are not needed anymore after building by limiting them to their own stage. For example, to compile `xar` we need to have a complete compiler collection installed (hundreds of MB), whereas to actually install it we only need `make` (a couple MB). The final image created by this is indeed only about 180MB larger than its' base, most of which is because of Java. With Java disabled it's only 14MB. 

Finally, when the image changes and circle (or any other client using it) pulls it, only the changed/new layers will need to be pulled, so the more layers can be cached the better. 

Which also reminds me, I should move the `COPY` lines to the end. This means that if ITMS or XAR is updated all the dependencies won't have to be reinstalled